### ISSUE TITLE
Memory leak in `Signal#discrete`

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#799]
This program should run in constant memory but it doesn't:

```scala
object SignalDiscreteSanityTest extends App {
  import TestUtil.S
  Stream.eval(async.signalOf[Task, Unit](())).flatMap { signal =>
    signal.discrete.evalMap(a => signal.set(a))
  }.run.unsafeRun
}
```

Based on some quick heap analysis, it looks like we are leaking resource cleanup -- e.g., a huge `LongMap`, a ton of `Token`s, etc.

Each element through `discrete` results in a new `Stream.bracket`. I suspect the finalizers are getting promoted to higher scope due to the async effect of `signal.set`.

Note: same test using `continuous` runs in constant memory:
```scala
object SignalContinuousSanityTest extends App {
  import TestUtil.S
  Stream.eval(async.signalOf[Task, Unit](())).flatMap { signal =>
    signal.continuous.evalMap(a => signal.set(a))
  }.run.unsafeRun
}
```

**Comment by mpilquist** [comment]
BTW, I'm pretty sure this is the cause of the memory leak in `time.awakeEvery` reported by @AusCoder in [Gitter](https://gitter.im/functional-streams-for-scala/fs2?at=589a98bf872fc8ce6208f34b)

**Comment by pchlupacek** [comment]
https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/core/shared/src/main/scala/fs2/async/mutable/Signal.scala#L106 is definitelly recursing the bracket pilling up cleanups

**Comment by pchlupacek** [comment]
I think we will need to create proper Tag for the discrete subscriber and cleanup registration from the outer scope of go.

**Comment by pchiusano** [comment]
@pchlupacek yup we just spotted the same thing... figuring out fix

**Comment by pchlupacek** [comment]
@pchiusano @mpilquist just for curiosity any progress on that? Eventually, I can tackle this, as we are about to move part of our production to fs2, with heavy signal usage. lmk pls. I would like also make 9.3 with this patch, or if there is something else worth of backporting from 1.0 lmk

**Comment by mpilquist** [comment]
I haven't made any progress (was busy with work stuff yesterday). I definitely agree we need a 0.9.3 with this fix. I'll look through the 1.0 commits and see what can be backported.

**Comment by pchlupacek** [comment]
lmk if you want to fix that or I shall do it. My approach will be to introduce Id/Tag of discrete subscriber and clean it on bracket finalisation, instead on every step. Any thoughts?

**Comment by mpilquist** [comment]
Sure, have at it.

**Comment by pchiusano** [comment]
Yeah, go for it... sorry I also didn't get to this yesterday.

**Comment by pchlupacek** [comment]
@mpilquist, @pchiusano see #802 with backport in #803

**Comment by pchlupacek** [comment]
also threw in Traverse improvements, hope you won't mind

**Comment by mpilquist** [comment]
Fixed in 0.9.3 and 1.0.0.
